core/trino-main/src/main/java/io/trino/sql/planner/ExpressionInterpreter.java

```diff
@@ -1130,22 +1132,30 @@ protected Object visitLikePredicate(LikePredicate node, Object context)
 
             // if pattern is a constant without % or _ replace with a comparison
             if (pattern instanceof Slice && (escape == null || escape instanceof Slice) && !isLikePattern((Slice) pattern, Optional.ofNullable((Slice) escape))) {
-                Slice unescapedPattern = unescapeLiteralLikePattern((Slice) pattern, Optional.ofNullable((Slice) escape));
                 Type valueType = type(node.getValue());
-                Type patternType = createVarcharType(unescapedPattern.length());
-                Type superType = typeCoercion.getCommonSuperType(valueType, patternType)
-                        .orElseThrow(() -> new IllegalArgumentException("Missing super type when optimizing " + node));
-                Expression valueExpression = toExpression(value, valueType);
-                if (!valueType.equals(superType)) {
-                    valueExpression = new Cast(valueExpression, toSqlType(superType), false, typeCoercion.isTypeOnlyCoercion(valueType, superType));
-                }
+                Slice unescapedPattern = unescapeLiteralLikePattern((Slice) pattern, Optional.ofNullable((Slice) escape));
+                VarcharType patternType = createVarcharType(countCodePoints(unescapedPattern));
+
+                Expression valueExpression;
                 Expression patternExpression;
-                if (superType instanceof VarcharType) {
+                if (valueType instanceof CharType) {
+                    if (((CharType) valueType).getLength() != patternType.getBoundedLength()) {
+                        return false;
+                    }
+                    valueExpression = toExpression(value, valueType);
+                    patternExpression = toExpression(trimTrailingSpaces(unescapedPattern), valueType);
+                }
+                else if (valueType instanceof VarcharType) {
+                    Type superType = typeCoercion.getCommonSuperType(valueType, patternType)
+                            .orElseThrow(() -> new IllegalArgumentException("Missing super type when optimizing " + node));
+                    valueExpression = toExpression(value, valueType);
+                    if (!valueType.equals(superType)) {
+                        valueExpression = new Cast(valueExpression, toSqlType(superType), false, typeCoercion.isTypeOnlyCoercion(valueType, superType));
```

Comment:
Why do we need to explicitly add `Cast` for `valueExpression`, but not for `unescapedPattern` below?

Reply:
same as in https://github.com/trinodb/trino/pull/8276#discussion_r651578182

requirements imposed by `TestExpressionInterpreter#assertOptimizedEquals` are kind of weird, but i didn't find an easy way to improve that